File: packages/ts-sdk-identity/src/verify.ts
import { decodeProtectedHeader, jwtVerify } from "jose";
import type { JWTVerifyOptions } from "jose";
import { getJwksKeyFn } from "./jwks.js";
import type { HimsJwtPayload, IdentityPluginOptions, Principal } from "./types.js";

const DEFAULT_MAX_TOKEN_AGE_SECONDS = 300;
const MAX_ALLOWED_TOKEN_AGE_SECONDS = 900;
const DEFAULT_CLOCK_SKEW_SECONDS = 60;
const MAX_CLOCK_SKEW_SECONDS = 60;
const DEFAULT_ALLOWED_ALGORITHMS = ["RS256"] as const;

export class IdentityVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IdentityVerificationError";
  }
}

function sanitizeRoles(raw: unknown): string[] {
  if (!Array.isArray(raw)) {
    throw new IdentityVerificationError("roles claim must be an array");
  }
  const normalized = raw
    .map((role) => (typeof role === "string" ? role.trim().toLowerCase() : ""))
    .filter((role) => role.length > 0);
  return [...new Set(normalized)].sort((a, b) => a.localeCompare(b));
}

function asNonEmptyString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new IdentityVerificationError(`${field} claim is required`);
  }
  return value.trim();
}

/**
 * Normalizes the optional `scopes` claim into a clean lowercase set. Non-arrays and non-string
 * members are dropped (never throws) — an absent/garbage `scopes` yields `[]`, i.e. a NON-operator
 * principal that still hard-requires a tenant. Only an explicit, signed `"platform"` member relaxes
 * the tenant requirement.
 */
function sanitizeScopes(raw: unknown): string[] {
  if (!Array.isArray(raw)) {
    return [];
  }
  const normalized = raw
    .map((scope) => (typeof scope === "string" ? scope.trim().toLowerCase() : ""))
    .filter((scope) => scope.length > 0);
  return [...new Set(normalized)].sort((a, b) => a.localeCompare(b));
}

/** True only for a bounded platform-operator token. Central so every relaxation shares one gate. */
function isPlatformScoped(scopes: readonly string[]): boolean {
  return scopes.includes("platform");
}

function normalizeAllowlist(
  value: string | string[],
  field: "issuer" | "audience",
): string[] {
  const values = Array.isArray(value) ? value : [value];
  if (values.length === 0) {
    throw new IdentityVerificationError(`${field} allowlist cannot be empty`);
  }

  const normalized = values
    .map((entry) => (typeof entry === "string" ? entry.trim() : ""))
    .filter((entry) => entry.length > 0);

  if (normalized.length === 0) {
    throw new IdentityVerificationError(`${field} allowlist cannot be empty`);
  }

  return [...new Set(normalized)];
}

function resolveMaxTokenAgeSeconds(options: IdentityPluginOptions): number {
  const maxTokenAgeSeconds = options.maxTokenAgeSeconds ?? DEFAULT_MAX_TOKEN_AGE_SECONDS;
  if (
    !Number.isFinite(maxTokenAgeSeconds) ||
    maxTokenAgeSeconds <= 0 ||
    maxTokenAgeSeconds > MAX_ALLOWED_TOKEN_AGE_SECONDS
  ) {
    throw new IdentityVerificationError(
      `maxTokenAgeSeconds must be within 1-${MAX_ALLOWED_TOKEN_AGE_SECONDS}`,
    );
  }
  return maxTokenAgeSeconds;
}

function resolveClockSkewSeconds(options: IdentityPluginOptions): number {
  const clockSkewSeconds = options.clockSkewSeconds ?? DEFAULT_CLOCK_SKEW_SECONDS;
  if (
    !Number.isFinite(clockSkewSeconds) ||
    clockSkewSeconds < 0 ||
    clockSkewSeconds > MAX_CLOCK_SKEW_SECONDS
  ) {
    throw new IdentityVerificationError(
      `clockSkewSeconds must be within 0-${MAX_CLOCK_SKEW_SECONDS}`,
    );
  }
  return clockSkewSeconds;
}

function resolveAllowedAlgorithms(options: IdentityPluginOptions): ReadonlySet<string> {
  const algorithms = options.allowedAlgorithms ?? DEFAULT_ALLOWED_ALGORITHMS;
  if (!Array.isArray(algorithms) || algorithms.length === 0) {
    throw new IdentityVerificationError("allowedAlgorithms must include at least one algorithm");
  }
  for (const alg of algorithms) {
    if (typeof alg !== "string" || alg.trim().length === 0) {
      throw new IdentityVerificationError("allowedAlgorithms contains invalid algorithm");
    }
  }
  return new Set(algorithms);
}

function validateProtectedHeader(
  token: string,
  allowedAlgorithms: ReadonlySet<string>,
): void {
  const header = decodeProtectedHeader(token);
  if (typeof header.kid !== "string" || header.kid.length === 0) {
    throw new IdentityVerificationError("JWT protected header missing kid");
  }
  if (typeof header.alg !== "string" || header.alg.length === 0) {
    throw new IdentityVerificationError("JWT protected header missing alg");
  }
  if (!allowedAlgorithms.has(header.alg)) {
    throw new IdentityVerificationError(`Unsupported JWT algorithm: ${header.alg}`);
  }
}

function resolveOrgIdForPrincipal(tenantId: string, raw: unknown): string {
  if (raw === null || raw === undefined) {
    return "";
  }
  if (typeof raw !== "string") {
    return "";
  }
  const trimmed = raw.trim();
  if (trimmed.length === 0) {
    return "";
  }
  if (trimmed === tenantId) {
    throw new IdentityVerificationError(
      "iq_tenant_id and org_id must represent distinct semantic scopes when org_id is set",
    );
  }
  return trimmed;
}

function toPrincipal(payload: HimsJwtPayload): Principal {
  const userId = asNonEmptyString(payload.sub, "sub");
  const scopes = sanitizeScopes(payload.scopes);
  // BET4 (security-critical): a bounded platform operator's token is tenant-LESS. Tolerate an
  // absent/empty `iq_tenant_id` ONLY when the signed token carries `scopes:["platform"]`; every
  // other token keeps the hard requirement. `scopes` is only issued from `platform_admins`
  // membership on an RS256-signed JWT (see identity-jwt-claims / create-hims-better-auth), so a
  // tenant user cannot forge this relaxation. When tenant-less, `tenantId` is "" (no sentinel).
  const tenantId = isPlatformScoped(scopes)
    ? typeof payload.iq_tenant_id === "string"
      ? payload.iq_tenant_id.trim()
      : ""
    : asNonEmptyString(payload.iq_tenant_id, "iq_tenant_id");
  const orgId = resolveOrgIdForPrincipal(tenantId, payload.org_id);
  const iss = asNonEmptyString(payload.iss, "iss");
  if (typeof payload.iat !== "number") {
    throw new IdentityVerificationError("iat claim is required");
  }
  if (typeof payload.exp !== "number") {
    throw new IdentityVerificationError("exp claim is required");
  }

  const sessionId =
    typeof payload.session_id === "string" && payload.session_id.trim().length > 0
      ? payload.session_id.trim()
      : "";

  return {
    userId,
    tenantId,
    orgId,
    roles: sanitizeRoles(payload.roles),
    scopes,
    sessionId,
    department: payload.department,
    iat: payload.iat,
    exp: payload.exp,
    iss,
  };
}

export async function verifyToken(
  token: string,
  options: IdentityPluginOptions,
): Promise<Principal> {
  const issuerAllowlist = normalizeAllowlist(options.issuer, "issuer");
  const audienceAllowlist = normalizeAllowlist(options.audience, "audience");

  const allowedAlgorithms = resolveAllowedAlgorithms(options);
  validateProtectedHeader(token, allowedAlgorithms);

  const keyFn = getJwksKeyFn(options.jwksUrl, options.cacheTtlMs);
  const maxTokenAgeSeconds = resolveMaxTokenAgeSeconds(options);
  const clockSkewSeconds = resolveClockSkewSeconds(options);

  const verifyOpts: JWTVerifyOptions = {};
  verifyOpts.issuer = issuerAllowlist;
  verifyOpts.audience = audienceAllowlist;
  verifyOpts.algorithms = [...allowedAlgorithms];
  // session_id is intentionally NOT a required claim:
  // - better-auth issues short-lived JWTs identified by `jti`; session state lives server-side
  //   in the `auth.session` table, not inside the access token.
  // - Resource services validate *identity* (sub, tenant, roles) — not auth-provider session
  //   lifecycle. Coupling to session_id would make every service a session-state consumer.
  // - If session-binding is needed later, it should be enforced at the auth gateway, not here.
  // `iq_tenant_id` is intentionally NOT jose-required: bounded platform-operator tokens are
  // tenant-less (BET4). The tenant requirement is enforced in `toPrincipal`, which hard-requires a
  // non-empty tenant for every token EXCEPT `scopes:["platform"]`. A non-operator token missing
  // `iq_tenant_id` therefore still fails (in toPrincipal), just with a clearer error than jose's.
  verifyOpts.requiredClaims = ["sub", "roles", "jti", "exp", "iat"];
  verifyOpts.maxTokenAge = `${maxTokenAgeSeconds}s`;
  verifyOpts.clockTolerance = `${clockSkewSeconds}s`;

  let payload: HimsJwtPayload;
  try {
    const verified = await jwtVerify<HimsJwtPayload>(token, keyFn, verifyOpts);
    payload = verified.payload;
  } catch (error) {
    if (
      error instanceof Error &&
      (error.name === "JWKSNoMatchingKey" || error.name === "JWKSInvalid")
    ) {
      throw new IdentityVerificationError(error.message);
    }
    throw error;
  }

  return toPrincipal(payload);
}

File: packages/py-sdk-authz/src/hims_authz/verify.py
"""In-process JWT verification (RS256 against the edge JWKS).

Python mirror of ``@hims/ts-sdk-identity`` ``verifyToken``. This is what actually closes
the direct-to-service bypass: a caller reaching opd/master-data directly must present a
JWT that verifies against the same JWKS the edge uses — an absent or forged token is
rejected here, in-process, before any handler runs.

Canonical tooling: PyJWT ``PyJWKClient`` (per-``kid`` cached, auto-refresh on unknown kid)
+ ``jwt.decode`` with RS256/audience/issuer/leeway. PyJWT has no ``max_age`` option, so the
``maxTokenAge`` bound is enforced with an explicit ``iat`` check (matching the TS side).
"""

from __future__ import annotations

import time
from collections.abc import Callable

import jwt
from jwt import PyJWK, PyJWKClient

from hims_authz.types import IdentityVerificationError, VerifiedIdentity

DEFAULT_MAX_TOKEN_AGE_SECONDS = 300
MAX_ALLOWED_TOKEN_AGE_SECONDS = 900
DEFAULT_CLOCK_SKEW_SECONDS = 60
MAX_CLOCK_SKEW_SECONDS = 60
ALLOWED_ALGORITHMS = ("RS256",)
# Mirrors verify.ts: session_id is intentionally NOT required (better-auth uses jti +
# short-lived tokens; session state is server-side).
REQUIRED_CLAIMS = ("sub", "iq_tenant_id", "roles", "jti", "exp", "iat")

# token -> verification key (PyJWK or a raw key). Injectable so tests can supply a key
# directly without standing up a JWKS HTTP endpoint.
SigningKeyResolver = Callable[[str], object]


def _normalize_allowlist(value: str | tuple[str, ...], field: str) -> list[str]:
    values = [value] if isinstance(value, str) else list(value)
    normalized = [v.strip() for v in values if isinstance(v, str) and v.strip()]
    if not normalized:
        raise IdentityVerificationError(f"{field} allowlist cannot be empty")
    # dedupe, preserve order
    return list(dict.fromkeys(normalized))


def _resolve_bounded(value: int, default: int, lo: int, hi: int, field: str) -> int:
    resolved = default if value is None else value
    if not isinstance(resolved, int) or resolved < lo or resolved > hi:
        raise IdentityVerificationError(f"{field} must be within {lo}-{hi}")
    return resolved


def _sanitize_roles(raw: object) -> tuple[str, ...]:
    if not isinstance(raw, list):
        raise IdentityVerificationError("roles claim must be an array")
    normalized = {
        r.strip().lower() for r in raw if isinstance(r, str) and r.strip()
    }
    return tuple(sorted(normalized))


def _require_non_empty_str(value: object, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise IdentityVerificationError(f"{field} claim is required")
    return value.strip()


def _resolve_org_id(tenant_id: str, raw: object) -> str:
    if not isinstance(raw, str):
        return ""
    trimmed = raw.strip()
    if not trimmed:
        return ""
    if trimmed == tenant_id:
        raise IdentityVerificationError(
            "iq_tenant_id and org_id must represent distinct scopes when org_id is set"
        )
    return trimmed


class TokenVerifier:
    """Verifies an access token and returns a :class:`VerifiedIdentity`.

    Construct once per process and reuse — ``PyJWKClient`` caches signing keys internally.
    """

    def __init__(
        self,
        *,
        jwks_url: str | None = None,
        issuer: str | tuple[str, ...],
        audience: str | tuple[str, ...],
        max_token_age_seconds: int = DEFAULT_MAX_TOKEN_AGE_SECONDS,
        clock_skew_seconds: int = DEFAULT_CLOCK_SKEW_SECONDS,
        signing_key_resolver: SigningKeyResolver | None = None,
        clock: Callable[[], float] = time.time,
    ) -> None:
        self._issuers = _normalize_allowlist(issuer, "issuer")
        self._audiences = _normalize_allowlist(audience, "audience")
        self._max_age = _resolve_bounded(
            max_token_age_seconds, DEFAULT_MAX_TOKEN_AGE_SECONDS,
            1, MAX_ALLOWED_TOKEN_AGE_SECONDS, "max_token_age_seconds",
        )
        self._skew = _resolve_bounded(
            clock_skew_seconds, DEFAULT_CLOCK_SKEW_SECONDS,
            0, MAX_CLOCK_SKEW_SECONDS, "clock_skew_seconds",
        )
        self._clock = clock

        if signing_key_resolver is not None:
            self._resolve_key = signing_key_resolver
        elif jwks_url:
            client = PyJWKClient(jwks_url)

            def _from_jwks(token: str) -> object:
                signing_key: PyJWK = client.get_signing_key_from_jwt(token)
                return signing_key.key

            self._resolve_key = _from_jwks
        else:
            raise IdentityVerificationError("jwks_url or signing_key_resolver is required")

    def verify(self, token: str) -> VerifiedIdentity:
        """Return the verified identity, or raise :class:`IdentityVerificationError`."""
        self._assert_header(token)

        try:
            key = self._resolve_key(token)
        except jwt.PyJWTError as exc:
            raise IdentityVerificationError(f"signing key resolution failed: {exc}") from exc

        try:
            payload = jwt.decode(
                token,
                key,
                algorithms=list(ALLOWED_ALGORITHMS),
                audience=self._audiences,
                issuer=self._issuers,
                leeway=self._skew,
                options={"require": list(REQUIRED_CLAIMS)},
            )
        except jwt.PyJWTError as exc:
            raise IdentityVerificationError(str(exc)) from exc

        self._assert_token_age(payload.get("iat"))
        return self._to_identity(payload)

    def _assert_header(self, token: str) -> None:
        try:
            header = jwt.get_unverified_header(token)
        except jwt.PyJWTError as exc:
            raise IdentityVerificationError(f"malformed token header: {exc}") from exc
        kid = header.get("kid")
        alg = header.get("alg")
        if not isinstance(kid, str) or not kid:
            raise IdentityVerificationError("JWT protected header missing kid")
        if not isinstance(alg, str) or alg not in ALLOWED_ALGORITHMS:
            raise IdentityVerificationError(f"unsupported JWT algorithm: {alg}")

    def _assert_token_age(self, iat: object) -> None:
        if not isinstance(iat, (int, float)):
            raise IdentityVerificationError("iat claim is required")
        now = self._clock()
        if iat > now + self._skew:
            raise IdentityVerificationError("iat is in the future")
        if now - iat > self._max_age + self._skew:
            raise IdentityVerificationError("token exceeds maximum age")

    def _to_identity(self, payload: dict[str, object]) -> VerifiedIdentity:
        user_id = _require_non_empty_str(payload.get("sub"), "sub")
        tenant_id = _require_non_empty_str(payload.get("iq_tenant_id"), "iq_tenant_id")
        iss = _require_non_empty_str(payload.get("iss"), "iss")
        jti = _require_non_empty_str(payload.get("jti"), "jti")
        org_id = _resolve_org_id(tenant_id, payload.get("org_id"))

        iat = payload.get("iat")
        exp = payload.get("exp")
        if not isinstance(iat, (int, float)):
            raise IdentityVerificationError("iat claim is required")
        if not isinstance(exp, (int, float)):
            raise IdentityVerificationError("exp claim is required")

        department_raw = payload.get("department")
        department = (
            department_raw.strip()
            if isinstance(department_raw, str) and department_raw.strip()
            else None
        )
        session_raw = payload.get("session_id")
        session_id = (
            session_raw.strip()
            if isinstance(session_raw, str) and session_raw.strip()
            else ""
        )

        return VerifiedIdentity(
            user_id=user_id,
            tenant_id=tenant_id,
            org_id=org_id,
            roles=_sanitize_roles(payload.get("roles")),
            department=department,
            session_id=session_id,
            jti=jti,
            iat=int(iat),
            exp=int(exp),
            iss=iss,
        )
